src/Adapter/FilesystemItems.php

```diff
@@ -0,0 +1,209 @@
+        } catch (ItemNotFound $e) {
+            $next = ItemVersionNumber::fromInt(1);
+        }
+
+        if ($version > $next) {
+            throw new UnexpectedVersionNumber($id, $version, $next);
+        }
+
+        $this->filesystem->dumpFile("{$this->path}/{$id}/{$version->toInt()}.xml", $item->getContent());
+    }
+
+    public function remove(ItemId $id, ?ItemVersionNumber $version) : void
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            return;
+        }
+
+        if (null === $version) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        try {
+            $highest = $this->get($id)->getVersion();
+        } catch (ItemNotFound $e) {
+            return;
+        }
+
+        if ($version->toInt() < $highest->toInt()) {
+            throw new UnexpectedVersionNumber($id, $version, $highest);
+        }
+
+        if (1 === $version->toInt()) {
+            $this->filesystem->remove("{$this->path}/{$id}");
+
+            return;
+        }
+
+        $this->filesystem->remove("{$this->path}/{$id}/{$version->toInt()}.xml");
+    }
+
+    public function get(ItemId $id, ?ItemVersionNumber $version = null) : ItemVersion
+    {
+        if (!is_dir("{$this->path}/{$id}")) {
+            throw new ItemNotFound($id);
+        }
+
+        if (null === $version) {
+            $versions = iterator_to_array($this->getVersions($id));
+            rsort($versions);
+            $version = $versions[0];
+        }
+
+        if (!$content = @fopen($file = "{$this->path}/{$id}/{$version->toInt()}.xml", 'rb')) {
+            throw new VersionNotFound($id, $version);
+        }
+
+        if (!$hash = md5_file($file)) {
+            throw new RuntimeException("Failed to hash {$file}");
+        }
+
+        return new ItemVersion($id, $version, $content, $hash);
+    }
+
+    public function list(int $limit = 10, ?ItemId $startAt = null) : ItemListPage
+    {
+        $ids = $this->getIds();
+
+        if (null === $startAt) {
+            $ids = new LimitIterator($ids, 0, $limit + 1);
+        } else {
+            $found = null;
+            $i = 0;
+            foreach ($ids as $id) {
+                if ($id == $startAt) {
+                    $found = $i;
+                    break;
+                }
+                $i++;
+            }
+
+            if (null === $found) {
+                $ids = new EmptyIterator();
+            } else {
+                $ids = new LimitIterator($ids, $found, $limit + 1);
+            }
+        }
+
+        $ids = array_values(iterator_to_array($ids));
+
+        if (count($ids) > $limit) {
+            $nextId = array_pop($ids);
+        }
+
+        return new ItemListPage($ids, $nextId ?? null);
+    }
+
+    public function count() : int
+    {
+        return iterator_count($this->getIds());
+    }
+
+    public function getIterator() : Traversable
+    {
+        $items = new MapIterator(
```

Comment:
I'm can only assume `MapIterator` is lazy, hence `get()` and the corresponding file opening won't happen until `yield`